File: leetcode/C++/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> st;
        ListNode *f=head, *s=head, *r=nullptr, *t;
        while(f && f->next){
            f=f->next->next;
            t=s->next;
            if(r){
                s->next=r;
                r=s;
            }
            else{
                r=s;
                r->next=nullptr;
            }
            s=t;
        }
        if(f) s=s->next;
        for(; s!=nullptr&&r!=nullptr; s=s->next, r=r->next){
            if(r->val!=s->val) return false;
        }
        return true;
    }
};

File: leetcode/Python/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        s, f, r, t = head, head, head, None
        while f and f.next:
            f, t = f.next.next, s.next
            if r:
                s.next, r = r, s
            else:
                r, r.next = s, None
            s=t
        if f:
            s=s.next
        while s and r:
            if s.val != r.val:
                return False
            s, r = s.next, r.next
        return True